### PR TITLE
PowerPC JIT compiler.

### PR DESCRIPTION
**Pull request by nioroso-x3**
New pull with squashed commits.

### CODE REVIEW

File: src/crypto/CryptonightR_JIT.c [hunk 1]
```diff
@@ -117,7 +204,78 @@ int v4_generate_JIT_code(const struct V4_Instruction* code, v4_random_math_JIT_f
 	__builtin___clear_cache((char*)buf, (char*)JIT_code);
 
 	return 0;
+#elif defined(__PPC__) || defined(__PPC64__)
+    uint8_t* JIT_code = (uint8_t*) buf;
+    const uint8_t* JIT_code_end = JIT_code + buf_size;
+    static const uint8_t regN[] = {4,5,6,7,8,9,10,11,12};
+    static const uint8_t r0 = 0;
+
+    #if __BYTE_ORDER == __BIG_ENDIAN
```
Comment (moneromooo-monero) [inline src/crypto/CryptonightR_JIT.c]: Is that PPC specific ? Monero uses BYTE_ORDER and BIG_ENDIAN.

File: src/crypto/slow-hash.c [hunk 2]
```diff
@@ -771,24 +812,23 @@ void slow_hash_allocate_state(void)
...
+      #define RESERVED_FLAGS 0
+    #endif
     hp_jitfunc_memory = mmap(0, 4096 + 4096, PROT_READ | PROT_WRITE | RESERVED_FLAGS,
                     MAP_PRIVATE | MAP_ANON, -1, 0);
-#else
+  #else
     hp_jitfunc_memory = mmap(0, 4096 + 4096, PROT_READ | PROT_WRITE | PROT_EXEC,
                     MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
-#endif
+  #endif
```
Comment (moneromooo-monero) [inline src/crypto/slow-hash.c]: Can you avoid all the reformatting please ?

File: src/crypto/slow-hash.c [hunk 3]
```diff
@@ -1579,20 +1618,213 @@ void cn_slow_hash(const void *data, size_t length, char *hash, int variant, int
 }
 #endif /* !aarch64 || !crypto */
 
+#elif defined(__PPC__) || (__PPC64__)
```
Comment (moneromooo-monero) [inline src/crypto/slow-hash.c]: Missing "defined" on the second test

File: src/crypto/slow-hash.c [hunk 4]
```diff
@@ -1579,20 +1618,213 @@ void cn_slow_hash(const void *data, size_t length, char *hash, int variant, int
...
+#include <sys/mman.h>
+//#include <SSE2ALTIVEC.h>
+// PPC64 implementation just a copy of the portable one for now, plus the JIT.
```
Comment (moneromooo-monero) [inline src/crypto/slow-hash.c]: Why is it necessary to duplicate all this stuff ?

File: src/crypto/slow-hash.c [hunk 5]
```diff
@@ -1579,20 +1618,213 @@ void cn_slow_hash(const void *data, size_t length, char *hash, int variant, int
...
 void slow_hash_allocate_state(void)
 {
-  // Do nothing, this is just to maintain compatibility with the upgraded slow-hash.c
+
   return;
 }
 
 void slow_hash_free_state(void)
 {
-  // As above
```
Comment (moneromooo-monero) [inline src/crypto/slow-hash.c]: Various gratuitous unrelated changes

**Comment by moneromooo-monero** [comment]
Looks like it gives the wrong hashes on the G5.

**Comment by moneromooo-monero** [comment]
I was going to test on the G5 to see if that can go in, but the server ssh key either changed, or someone's MITM the connection. Did it change ?

**Comment by hyc** [comment]
Since the last status on this PR is "don't merge this" and CNv4 is about to become obsolete anyway, perhaps this should be closed.

**Comment by selsta** [comment]
@hyc Can you close this? You should have the rights for this repo.
